### PR TITLE
test: additional tests for dynamic arrays

### LINKED ISSUE #2594: add more tests for dynamic arrays
A few that come to mind

- [ ] test clamps for nested dynamic arrays
- [ ] test array access for nested dynamic arrays
- [ ] test various kinds of nesting, e.g. tuples of dynamic arrays or dynamic arrays of tuples
- [ ] test multiple levels of nesting
- [ ] test immutable (and nested) dynamic arrays
- [ ] test nested dynamic arrays in constructor

### PR DESCRIPTION
**Pull request by tserg**
### What I did

Closes #2594 

Add tests for dynamic arrays.

### How I did it

Go through list of tasks

### How to verify it

Run tests

### Description for the changelog

Additional tests for dynamic arrays

### Cute Animal Picture

Put a link to a cute animal picture inside the parenthesis-->

### CODE REVIEW

File: tests/functional/codegen/test_abi_encode.py [hunk 1]
```diff
@@ -134,6 +134,100 @@ def foo():
...
+        if not include_method_id:
+            return _abi_encode(d, ensure_tuple=False)
+        return _abi_encode(d, ensure_tuple=False, method_id=0xdeadbeef)
+    """
+    c = get_contract(code)
+
+    method_id = 0xDEADBEEF .to_bytes(4, "big")
+
+    arg = [
+        [[123, 456, 789], [234, 567, 891], [345, 678, 912]],
```
Comment (charles-cooper) [inline tests/functional/codegen/test_abi_encode.py:216]: i think it would be good to have some test data which is partially empty

Comment (charles-cooper) [reply tests/functional/codegen/test_abi_encode.py:216]: i think important cases are: empty, one-less-than full, full

Comment (tserg) [reply tests/functional/codegen/test_abi_encode.py]: Got it, have added other test data.

**Review by charles-cooper — APPROVED** [review]
would be good to get some DynArray of string tests in there but we can get those in a follow up PR

File: tests/functional/codegen/test_abi_encode.py [hunk 2]
```diff
@@ -134,6 +134,163 @@ def foo():
...
+            return _abi_encode(d) # default ensure_tuple=True
+        return _abi_encode(d, method_id=0xdeadbeef)
+    else:
+        if not include_method_id:
+            return _abi_encode(d, ensure_tuple=False)
+        return _abi_encode(d, ensure_tuple=False, method_id=0xdeadbeef)
+    """
+    c = get_contract(code)
+
+    method_id = 0xDEADBEEF .to_bytes(4, "big")
```
Comment (fubuloubu) [inline tests/functional/codegen/test_abi_encode.py:152]:
```suggestion
    method_id = 0xDEADBEEF.to_bytes(4, "big")
```

**Review by fubuloubu — APPROVED** [review]

**Pull request merged**